### PR TITLE
HTLC upstream close debugging additions

### PR DESCRIPTION
**Pull request by rustyrussell**
Firstly, I added an explicit test to try to mirror https://github.com/ElementsProject/lightning/issues/4649

Then I added more debugging and checks if it happens again.

### CODE REVIEW

File: tests/test_closing.py [hunk 1]
```diff
@@ -3631,3 +3631,71 @@ def test_close_weight_estimate(node_factory, bitcoind):
...
+    l1.rpc.sendpay(route, ph2)
+
+    # This can take 30 seconds...
+    l2.daemon.wait_for_log('Adding HTLC 1 too slow: killing connection',
+                           timeout=TIMEOUT + 30)
+    l2.daemon.wait_for_log('Failing HTLC 1 due to peer death')
+
+    with pytest.raises(RpcError, match=r'WIRE_TEMPORARY_CHANNEL_FAILURE \(reply from remote\)'):
+        l1.rpc.waitsendpay(ph2, timeout=TIMEOUT)
+
```
Comment (vincenzopalazzo) [inline tests/test_closing.py:3672]: There should be extra whitespace that I don't see here.

A simple space to stop the CI :/

File: tests/test_closing.py [hunk 2]
```diff
@@ -3631,3 +3631,41 @@ def test_close_weight_estimate(node_factory, bitcoind):
     signed_weight = int(bitcoind.rpc.decoderawtransaction(tx)['weight'])
     assert signed_weight <= actual_weight
     assert signed_weight >= actual_weight - 2
+
+
+@pytest.mark.developer("needs dev_disconnect")
+def test_onchain_close_upstream(node_factory, bitcoind, executor):
+    """We send an HTLC, and peer unilaterally closes: do we close
+upstream?  Aka
+https://github.com/ElementsProject/lightning/issues/4649
```
Comment (cdecker) [inline tests/test_closing.py:3640]: Nit: Indentation and first line should be short enough not to wrap, followed by an empty line with the long description.

**Review by cdecker — COMMENTED** [review]
Seems this is indeed surfacing an issue here:

```
2022-03-28T01:07:06.1316726Z lightningd-3: 2022-03-28T00:17:27.274Z **BROKEN** 022d223620a359a47ff7f7ac447c85c46c923da53389221a0054c11c1e3ca31d59-chan#1: Internal error ONCHAIN: onchaind_missing_htlc_output: htlc 4 is not local!
```

File: tests/test_closing.py (hunk 2 above)
Comment (rustyrussell) [reply tests/test_closing.py]: Fixed, thanks!

**Comment by rustyrussell** [comment]
> Seems this is indeed surfacing an issue here:
> 
> ```
> 2022-03-28T01:07:06.1316726Z lightningd-3: 2022-03-28T00:17:27.274Z **BROKEN** 022d223620a359a47ff7f7ac447c85c46c923da53389221a0054c11c1e3ca31d59-chan#1: Internal error ONCHAIN: onchaind_missing_htlc_output: htlc 4 is not local!
> ```

And, indeed, this lead me to the smoking gun!  Phew....

**Review by cdecker — APPROVED** [review]
ACK 0a119301c8033fa77cdffb68b798137b52a33b96

**Comment by cdecker** [comment]
Seems there is just a memleak that needs to be cleaned up, otherwise gtg :-)

**Comment by rustyrussell** [comment]
> Seems there is just a memleak that needs to be cleaned up, otherwise gtg :-)

Gah, allocated temporary off ctx not tmpctx.  Trivial fix...

**Pull request merged**